prometheus-xmpp-alerts

```diff
@@ -65,8 +65,8 @@ class XmppApp(slixmpp.ClientXMPP):
         self.get_roster()
 
     def lost(self, event):
-        logging.info("Connection lost, exiting.")
-        sys.exit(1)
+        logging.info("Reconnecting...")
+        self.connect()
```

Comment:
The current behaviour is intentional - I think it's generally better to have the parent restart the process when the XMPP connection is dropped.

At the very least, if this reconnects it should have a limited number of retries and perhaps sleep in between retries to avoid dossing the server.

Reply:
@jelmer no objection to that, but in my setup I get frequent `INFO     connection_lost: (ConnectionResetError(104, 'Connection reset by peer'),)` which I still have to debug.
They lead to prometheus alerting me about the xmpp-alert service flapping, so I preferred this fix. (Also, I do `pip install --upgrade` in `ExecStartPre` of my systemd unit, which would be run unnecessarily often)

Reply:
Okay, i'll look into adding an back-off period between reconnects, but i'd generally prefer to have the process stay up as long as possible - also to keep the metrics from being reset, just because of a disconnect! Maybe even make it retain its session identifier.

Reply:
(In the meantime you could maybe already merge #14, if you're fine with allowing the verbose message format as just an opt-in.)

Reply:
I've put some time into analyzing this, and it seems as if no further "optimization" of the reconnects is needed.
I came to this conclusion because the `disconnected` event is only fired once - when the client is disconnected. The following reconnect attempts are then handled by SLIXMPP, so it's not like we're doing a forced loop of `self.connect()` then, because there has been no additional `disconnected` event... as we haven't been connected since the last event.
So by killing and consequently restarting the parent process, thereby forcing an immediate reconnect, you're probably using a more aggressive reconnection strategy than by letting the client just do its thing, as it seems as if there's some back-off logic already in place.

Reply:
What if the server is e.g. full and refuses the connection? With the current code, it would attempt to connect over and over again indefinitely, without waiting.

E.g. systemd does back-off; if the service fails to start repeatedly, it'll be disabled or restarts can be delayed. At the very least, a similar mechanism is necessary here.

Reply:
To state it once more explicitly: the reconnect happens **only** after a `disconnected` event.
If your client wasn't able to connect because the server refuses the connection, then it wasn't connected, and so there can't be a `disconnected` event unless it was able successfully connect.

And as i said, i'm pretty sure i noticed an increase in the connection intervals - for failed connection attempts, which is a completely different event - without any code modification when running in debug mode. That seems to be handle by the module already. If you want to change that behaviour, you'd need to add a completely separate event handler for the `connection_failed` event.

... and to be honest: if a server can be DOS'ed by a single XMPP client trying to connect, you should switch to a different server. :grin:

Reply:
The fact that it won't bring the server doesn't make it good practice.

It may still get you banned if you're e.g. not running the server. It will also cause unnecessary extra CPU consumption on the client and spams its logs.

If you can point at behaviour in slixmpp that causes it to do exponential backoff in the case of disconnects, then perhaps this change would be more reasonable. But I am not aware of such behaviour. Anecdotally, I see it reconnecting immediately (i.e. every second or so) when I apply your patch.

I don't see how this change is an improvements over the current situation where systemd would just restart prometheus-xmpp-alerts. 

I think it would be reasonable to do the reconnect in prometheus-xmpp-alerts itself (rather than exiting and letting systemd take care of it), but in that case it should at the very least have some delay if it reconnects repeatedly and/or have exponential backoff.

Reply:
There you go: https://sleekxmpp.readthedocs.io/en/latest/api/xmlstream/xmlstream.html#sleekxmpp.xmlstream.xmlstream.XMLStream.connect

Reply:
So shouldn't we just set "reattempt=True" and not doing anything in the disconnect handler?

Reply:
The code you linked to is for sleekxmpp; slixmpp doesn't support a reattempt argument as far as I can tell and it just waits 2 seconds between reconnects by default - there is no exponential backoff.